### ISSUE TITLE
Explore the heavily changed Rasa NLU training progress

### ISSUE THREAD

**Issue by samsucik** [RasaHQ/rasa#8576]
**Rasa version**: 2.5.1 (current `main` branch)

**Python version**: 3.7.9

**Operating system** (windows, osx, ...): MacOS

**Issue**: 
After the recent DIET & TED refactoring PR (#7589) got merged, DIET's training characteristic (how the training loss and accuracy evolves) changed drastically -- at least for one small dataset we've got (see [this Slack thread](https://rasa-hq.slack.com/archives/C8LEHMTAM/p1619685661333200)).

This matters because convergence properties shouldn't change (degrade) this much, and the changes introduced in the PR were not meant to change this at all.

What needs to be done:
- [x] check if a similar effect can be observed on other NLU datasets
- [x] check what the impact is on the test-set results (because training accuracy/loss and generalisation performance are different things)
- [x] explore the different versions (commits) of Rasa OSS to pinpoint the changes that impacted this
- [ ] identify what (if anything) needs to change in the code if we want to "bring things back to normal"

**Error**:
![plot](https://user-images.githubusercontent.com/7470182/116589631-be2d1e00-a914-11eb-8400-7f4bbe36e004.png)

**Content of configuration file (config.yml)** (if relevant):
```yml
language: en
pipeline:
  - name: WhitespaceTokenizer
  - name: RegexFeaturizer
  - name: LexicalSyntacticFeaturizer
  - name: CountVectorsFeaturizer
  - name: CountVectorsFeaturizer
    analyzer: "char_wb"
    min_ngram: 1
    max_ngram: 4
  - name: DIETClassifier
    epochs: 100
    random_seed: 47
```

**Comment by samsucik** [comment]
Something weird happens to training accuracy very early in the history of changes on the PR (#7589) and its predecessor (#7378). I went through all the commits on these PRs and pinned down the consecutive commit pairs where the numbers change. Below are results that should cover the key commits and some more.

For now ignoring the test-set accuracy, here are stats for the final training accuracy (what one sees in the training progress bar); averaged over the same 4 random seeds, with std. deviation shown as well. Ordered by commit time, from oldest to newest. All the commits come from the 2 mentioned PRs.

The "car bot":
```
timestamp_commit-hash | datetime | final train acc. ± stdev | test acc. ± stdev
1605195798_f87c134f7aed137944dbf3be49cb9da30bbeceee | 2020-11-12 15:43:18 | 0.95±0.02 | 0.00±0.00
1606814828_537d50b380d514842734ef74dbd250d8d89ca9dd | 2020-12-01 09:27:08 | 0.95±0.02 | 0.00±0.00
1607508327_63af6f2cb5d463036116493e647cc63d3b8c2240 | 2020-12-09 10:05:27 | 0.72±0.04 | 0.00±0.00
1607616261_06d2d7df71aac51db7325e4dd0ad521c3652ea0a | 2020-12-10 16:04:21 | 0.95±0.02 | 0.00±0.00
1607625787_9fe46d12723541dd2902caa249763faaf105e38c | 2020-12-10 18:43:07 | 0.95±0.02 | 0.00±0.00
1607626469_7705ee1b5ac87d87761f5f6f09ca7dfa504f01b9 | 2020-12-10 18:54:29 | 0.95±0.02 | 0.00±0.00
1607943517_d2722ec839be661a027e5010e47cb728ba6bf7eb | 2020-12-14 10:58:37 | 0.95±0.02 | 0.00±0.00
1607950842_2021e7d8ae61669f7162bf4d11100e59a92c4eea | 2020-12-14 13:00:42 | 0.95±0.02 | 0.00±0.00
1607951395_09a52f0a826671cfb14a136279dedfcd02d55c5b | 2020-12-14 13:09:55 | 0.95±0.02 | 0.00±0.00
1607953218_dbf7e9726c0280707be59cf1e2fb975aa8cfb283 | 2020-12-14 13:40:18 | 0.72±0.04 | 0.00±0.00
1607955711_8540f37cf2a5a1aadba48384817dd58e7bc07476 | 2020-12-14 14:21:51 | 0.72±0.04 | 0.00±0.00
1608225738_83c2a07c7a87445c11fbf9bb2ab3f156c8582853 | 2020-12-17 17:22:18 | 0.70±0.04 | 0.00±0.00
1608225974_4ba9505214c13f074c6c62b9185f2a6073894d3f | 2020-12-17 17:26:14 | 0.70±0.04 | 0.00±0.00
1610447898_d6e9b1a665efe56a5cc79840c6e1b3b0a3b3845e | 2021-01-12 10:38:18 | 0.70±0.04 | 0.00±0.00
1610728213_042ae44113bbedb08664fc3bc0766f5c107c4bc2 | 2021-01-15 16:30:13 | 0.70±0.04 | 0.00±0.00
1611076524_ac51434766c1341afdd48b5480d3e8ecb69888a8 | 2021-01-19 17:15:24 | 0.70±0.04 | 0.00±0.00
1611235280_c65504244d31c317a7fa569fcd13c7e4f30b3e28 | 2021-01-21 13:21:20 | 0.70±0.04 | 0.00±0.00
1611760324_9dc3298868cba683f03d72017472d9d035dabc46 | 2021-01-27 15:12:04 | 0.74±0.04 | 0.00±0.00
1611764124_661d9a877086dc951c58f27cb3267ab757e88d86 | 2021-01-27 16:15:24 | 0.75±0.03 | 0.00±0.00
1611825375_a3a73fd19846071bcd848fcba8ba7c58a55896ac | 2021-01-28 09:16:15 | 0.75±0.03 | 0.00±0.00
1611910626_d8167f71f5ec9d559e9b97d0b63fadd288fe48ab | 2021-01-29 08:57:06 | 0.75±0.03 | 0.00±0.00
1611940328_41a84f4414604bf6844b27a73c7923ef2d15070c | 2021-01-29 17:12:08 | 0.75±0.03 | 0.00±0.00
1612201842_3701622db3a24832aaa7a935f011f1e0546aaa8b | 2021-02-01 17:50:42 | 0.77±0.02 | 0.00±0.00
1612201969_c4c144e04e45261bbe56cc183eb29862bc3ae619 | 2021-02-01 17:52:49 | 0.77±0.02 | 0.00±0.00
1612285910_ea3f6472acbaf492d6187ca0c4b4d0ea9bc5e118 | 2021-02-02 17:11:50 | 0.77±0.02 | 0.00±0.00
1614002994_08f9f1ebfb15eb1cc9fcafe5d4332dbb3daae005 | 2021-02-22 14:09:54 | 0.74±0.04 | 0.00±0.00
1615225828_aaaef4cb9dba90a768fa53fc7a6c84da82822c08 | 2021-03-08 17:50:28 | 0.74±0.04 | 0.00±0.00
1616162263_5df03ee3a51901901c9c002a5297ed8a9d4b8d74 | 2021-03-19 13:57:43 | 0.76±0.03 | 0.00±0.00
1618228839_fe864b6428bbbdf9fba95865aa73fa67f93c5aae | 2021-04-12 13:00:39 | 0.76±0.03 | 0.00±0.00
1619427287_8e2a8a060e7b1d2a9aad3004b7a2562c8ad79bd0 | 2021-04-26 09:54:47 | 0.76±0.03 | 0.00±0.00
```

And Carbon bot (numbers for some last commits aren't computed yet):
```
timestamp_commit-hash | datetime | final train acc. ± stdev | test acc. ± stdev
1605195798_f87c134f7aed137944dbf3be49cb9da30bbeceee | 2020-11-12 15:43:18 | 0.99±0.00 | 0.00±0.00
1606814828_537d50b380d514842734ef74dbd250d8d89ca9dd | 2020-12-01 09:27:08 | 0.99±0.00 | 0.00±0.00
1607508327_63af6f2cb5d463036116493e647cc63d3b8c2240 | 2020-12-09 10:05:27 | 0.82±0.01 | 0.00±0.00
1607616261_06d2d7df71aac51db7325e4dd0ad521c3652ea0a | 2020-12-10 16:04:21 | 0.99±0.00 | 0.00±0.00
1607625787_9fe46d12723541dd2902caa249763faaf105e38c | 2020-12-10 18:43:07 | 0.99±0.00 | 0.00±0.00
1607626469_7705ee1b5ac87d87761f5f6f09ca7dfa504f01b9 | 2020-12-10 18:54:29 | 0.99±0.00 | 0.00±0.00
1607943517_d2722ec839be661a027e5010e47cb728ba6bf7eb | 2020-12-14 10:58:37 | 0.99±0.00 | 0.00±0.00
1607950842_2021e7d8ae61669f7162bf4d11100e59a92c4eea | 2020-12-14 13:00:42 | 0.99±0.00 | 0.00±0.00
1607951395_09a52f0a826671cfb14a136279dedfcd02d55c5b | 2020-12-14 13:09:55 | 0.99±0.00 | 0.00±0.00
1607953218_dbf7e9726c0280707be59cf1e2fb975aa8cfb283 | 2020-12-14 13:40:18 | 0.82±0.01 | 0.00±0.00
1607955711_8540f37cf2a5a1aadba48384817dd58e7bc07476 | 2020-12-14 14:21:51 | 0.82±0.01 | 0.00±0.00
1608225738_83c2a07c7a87445c11fbf9bb2ab3f156c8582853 | 2020-12-17 17:22:18 | 0.81±0.00 | 0.00±0.00
1608225974_4ba9505214c13f074c6c62b9185f2a6073894d3f | 2020-12-17 17:26:14 | 0.81±0.00 | 0.00±0.00
1610447898_d6e9b1a665efe56a5cc79840c6e1b3b0a3b3845e | 2021-01-12 10:38:18 | 0.81±0.00 | 0.00±0.00
1610728213_042ae44113bbedb08664fc3bc0766f5c107c4bc2 | 2021-01-15 16:30:13 | 0.81±0.00 | 0.00±0.00
1611076524_ac51434766c1341afdd48b5480d3e8ecb69888a8 | 2021-01-19 17:15:24 | 0.81±0.00 | 0.00±0.00
1611235280_c65504244d31c317a7fa569fcd13c7e4f30b3e28 | 2021-01-21 13:21:20 | 0.81±0.00 | 0.00±0.00
1611760324_9dc3298868cba683f03d72017472d9d035dabc46 | 2021-01-27 15:12:04 | 0.81±0.01 | 0.00±0.00
1611764124_661d9a877086dc951c58f27cb3267ab757e88d86 | 2021-01-27 16:15:24 | 0.83±0.00 | 0.00±0.00
1611825375_a3a73fd19846071bcd848fcba8ba7c58a55896ac | 2021-01-28 09:16:15 | 0.83±0.00 | 0.00±0.00
1611910626_d8167f71f5ec9d559e9b97d0b63fadd288fe48ab | 2021-01-29 08:57:06 | 0.83±0.00 | 0.00±0.00
1611940328_41a84f4414604bf6844b27a73c7923ef2d15070c | 2021-01-29 17:12:08 | 0.83±0.00 | 0.00±0.00
1612201842_3701622db3a24832aaa7a935f011f1e0546aaa8b | 2021-02-01 17:50:42 | 0.77±0.07 | 0.00±0.00
1612201969_c4c144e04e45261bbe56cc183eb29862bc3ae619 | 2021-02-01 17:52:49 | 0.81±0.00 | 0.00±0.00
1612285910_ea3f6472acbaf492d6187ca0c4b4d0ea9bc5e118 | 2021-02-02 17:11:50 | 0.81±0.00 | 0.00±0.00
1614002994_08f9f1ebfb15eb1cc9fcafe5d4332dbb3daae005 | 2021-02-22 14:09:54 | 0.82±0.01 | 0.00±0.00
1615225828_aaaef4cb9dba90a768fa53fc7a6c84da82822c08 | 2021-03-08 17:50:28 | 0.82±0.01 | 0.00±0.00
1616162263_5df03ee3a51901901c9c002a5297ed8a9d4b8d74 | 2021-03-19 13:57:43 | 0.83±0.01 | 0.00±0.00
1618228839_fe864b6428bbbdf9fba95865aa73fa67f93c5aae | 2021-04-12 13:00:39 | 0.83±0.01 | 0.00±0.00
1619427287_8e2a8a060e7b1d2a9aad3004b7a2562c8ad79bd0 | 2021-04-26 09:54:47 | 0.83±0.01 | 0.00±0.00
```

More coming soon.

**Comment by samsucik** [comment]
@dakshvar22 found the bug, dropout is being applied also to label features when it should only be used for non-label ones.

**Comment by samsucik** [comment]
Having fixed the bug in the linked PR, I also checked TED and ResponseSelector to see what the training curves looked just before the buggy PR got merged into `main`, just afterwards, and after my fix. While the curve for ResponseSelector doesn't show any change (because the bug only affects the `use_text_as_label` mode, which is not used here), the curve for TED does. Interestingly, with the bug, training performance went up but test-set score (in parentheses in the legend) went down:
![plot-after-fix](https://user-images.githubusercontent.com/7470182/117169265-ddb6c180-adc0-11eb-9d30-b2f4c5e586bc.png)

I'm also working on producing a similar plot for Sara.

Note that the +- and stdev are misleading here, as it's all results from just one fixed random seed.
